### PR TITLE
Added DTO prototype

### PR DESCRIPTION
**Pull request by sober607**

### CODE REVIEW

File: Library.Business/DTO/Book/BookDto.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (IhnatenkoDenys) [inline Library.Business/DTO/Book/BookDto.cs:5]: remove unused usings

File: Library.Business/DTO/Country/UpdateCountryDto.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Library.Business.DTO.Country
+{
+    public class UpdateCountryDto
```
Comment (IhnatenkoDenys) [inline Library.Business/DTO/Country/UpdateCountryDto.cs:9]: Why are additional models needed? )
It's good to have a single model for creating and updating

File: Library.Business/DTO/Book/CreateBookDto.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Library.Business.DTO.Book
+{
+    public class CreateBookDto
```
Comment (IhnatenkoDenys) [inline Library.Business/DTO/Book/CreateBookDto.cs:9]: Why do we need two almost identical models?

File: Library.Business/Mapping/ModelMappingProfile.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        public ModelMappingProfile()
+        {
+            // To review and edit
+            CreateMap<Book, BookDto>().ForMember(destination => destination.AuthorIds, conf => conf.MapFrom(x => x.BookAuthors.Select(y => y.AuthorId).ToList())); // Проверитьб нужен ли ToLLIst
+            CreateMap<BookDto, Book>();
+
+            CreateMap<CreateBookDto, Book>();
+
+            CreateMap<Country, CountryDto>();
+            CreateMap<CountryDto, Country>();
```
Comment (IhnatenkoDenys) [inline Library.Business/Mapping/ModelMappingProfile.cs:22]: U can use
.ReverseMap();

File: Library.Business/DTO/Person/PersonDto.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+using System.Threading.Tasks;
+
+namespace Library.Business.DTO.Person
+{
+    public class PersonDto
+    {
+        public long Id { get; set; }
+
+        public string FirstName { get; set; }
+#nullable enable
```
Comment (IhnatenkoDenys) [inline Library.Business/DTO/Person/PersonDto.cs:14]: perhaps you should avoid directives (#nullable) and try to move such configuration to project settings if possible

File: Library.Business/Mapping/ModelMappingProfile.cs (hunk 4 above)
Comment (IhnatenkoDenys) [reply Library.Business/Mapping/ModelMappingProfile.cs:22]: Also, you can skip creating DTO models in the business layer.

File: Library.Business/Model/ResultModel/ResultModel.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        public ErrorData Error { get; set; }
+
+        public static ResultModel<T> GetSuccess(T transferredData)
+        {
+            if(object.Equals(transferredData, default(T)) && typeof (T) != typeof(bool))
+            {
+                throw new ArgumentNullException();
+            }
+            else
```
Comment (IhnatenkoDenys) [inline Library.Business/Model/ResultModel/ResultModel.cs:21]: U can avoid using else statement

File: Library.DAL/Repositories/CountryRepository.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class CountryRepository : EfCoreRepository<Country, ApplicationContext>, ICountryRepository
+    {
+        public CountryRepository(ApplicationContext context) : base(context)
+        {
+        }
+
+        public IQueryable<Country> GetCountriesByNameFragement(string countryNameFragment)
+        {
+             return _context.Countries.Where(x => x.Name.Contains(countryNameFragment)).AsQueryable();
```
Comment (IhnatenkoDenys) [inline Library.DAL/Repositories/CountryRepository.cs:18]: remove .AsQueryable() :)

File: Library.DAL/Repositories/CountryRepository.cs [hunk 8]
```diff
@@ -1,12 +1,21 @@
...
 namespace Library.DAL.Repositories
 {
-    public class CountryRepository : EfCoreRepository<Country, ApplicationContext>
+    public class CountryRepository : EfCoreRepository<Country, ApplicationContext>, ICountryRepository
     {
         public CountryRepository(ApplicationContext context) : base(context)
         {
         }
+
+        public IQueryable<Country> GetCountriesByNameFragement(string countryNameFragment)
```
Comment (IhnatenkoDenys) [inline Library.DAL/Repositories/CountryRepository.cs:16]: In this case, you can use IEnumarable
IQueryble is used for the non-constant "where" statement

Comment (IhnatenkoDenys) [reply Library.DAL/Repositories/CountryRepository.cs:16]: IQueryable<T> Find(Expression<Func<T, bool>> predicate = null);
IQueryable<T> FindAll();

public IQueryable<T> Find(Expression<Func<T, bool>> predicate = null)
{
    return this.Set.Where(predicate); // Set is DBSet
}

and somewhere in the code .ToList() or .FirstOrDefault(), etc.

File: Library.Business/DTO/Country/UpdateCountryDto.cs (hunk 2 above)
Comment (sober607) [reply Library.Business/DTO/Country/UpdateCountryDto.cs:9]: Its what tearchers at IT Academy told us. Separate DTO for each action, if fields which is necessary for action is different

File: Library.Business/DTO/Person/PersonDto.cs (hunk 5 above)
Comment (sober607) [reply Library.Business/DTO/Person/PersonDto.cs:14]: If I apply nullable for all project, every String, List will result a warning, which says I have to set this variables as nullable "?", but i dont want to do some strings nullable. 
Looks like this has to be still investigated.

File: Library.Business/Mapping/ModelMappingProfile.cs (hunk 4 above)
Comment (sober607) [reply Library.Business/Mapping/ModelMappingProfile.cs:22]: It has to be done at API layer?
My thought was - as Service methods are at Business, so it should be on same level

**Pull request closed without merging**